### ISSUE TITLE
Unexpected behaviour with getSubfieldsValues in update trait

### ISSUE THREAD

**Issue by fournodes** [Laravel-Backpack/CRUD#4387]
There appears to be an issue with function getSubfieldsValues in the update trait.
Specifically the part that handles fields that have a dot in the name

https://github.com/Laravel-Backpack/CRUD/blob/bdf1349c88f8f621f568c97f6cad16e26f788b62/src/app/Library/CrudPanel/Traits/Update.php#L255-L266

Let assume we have a model called **Address** with attribute _street_number_ having an *integer* value of **1** so
`$name = "address.street_number"`

Right after this block of code is executed
https://github.com/Laravel-Backpack/CRUD/blob/bdf1349c88f8f621f568c97f6cad16e26f788b62/src/app/Library/CrudPanel/Traits/Update.php#L260-L263
The iterator has the following value
`$iterator = 1`

***Is there a reason why the code assumes the attribute value would be a string if it is not a model instance ?***
https://github.com/Laravel-Backpack/CRUD/blob/bdf1349c88f8f621f568c97f6cad16e26f788b62/src/app/Library/CrudPanel/Traits/Update.php#L258-L259

Given that iterator now contains a number, as you can expect
```php
!is_string($iterator = 1) // true
&&
!is_null($iterator = 1) // true

$this->getModelWithFakes($iterator = 1)->getAttributes()
```
So when getModelWithFakes is called it results in an exception when class_uses_recursive is called using a value that is not a model
https://github.com/Laravel-Backpack/CRUD/blob/bdf1349c88f8f621f568c97f6cad16e26f788b62/src/app/Library/CrudPanel/Traits/Update.php#L195-L202

**Solution:**
If the idea is to call getModelWithFakes only in the case when $iterator is a model then perhaps the code could be rewritten as
```php 
$iterator instanceof \Illuminate\Database\Eloquent\Model ? $this->getModelWithFakes($iterator)->getAttributes() : $iterator
```

@pxpm

**Comment by pxpm** [comment]
Hello @fournodes 

Thanks for the report. Unfortunately I was not able to reproduce it, maybe I am using the wrong usecase. 

From my understanding some of your `subfields` are `dot.notation` fields. We have the same example in our demo: https://demo.backpackforlaravel.com 

In `Monsters & Stuff` menu, you will find the `Story` for example, this Story HasMany Monsters, and those Monsters have some fields inside that use `dot.notation` as for example `address.street`. 

Can you please provide me a more detailed implementation of your usecase so I can attempt to reproduce it ? 

Cheers

**Comment by fournodes** [comment]
Hey @pxpm

It is actually pretty simply to replicate the issue, simply modify the _street_ column in the **addresses** table to any non-string type like **integer**

I did that on MySQL
```mysql
ALTER TABLE `addresses` MODIFY `street` INTEGER;
```

Here is the before and after view from tinker
![image](https://user-images.githubusercontent.com/62574658/169597642-e62edc84-4278-4973-9076-acd4b3f725fd.png)

Follow that by opening edit for any story record
![image](https://user-images.githubusercontent.com/62574658/169598077-ead0cf11-3191-4de8-aab0-98a5b413c78a.png)

**Comment by fournodes** [comment]
Given now you should be able to replicate the issue, it would be much appreciated if you could answer this

> ***Is there a reason why the code assumes the attribute value would be a string if it is not a model instance ?***

https://github.com/Laravel-Backpack/CRUD/blob/bdf1349c88f8f621f568c97f6cad16e26f788b62/src/app/Library/CrudPanel/Traits/Update.php#L258-L259

**Comment by pxpm** [comment]
Thanks @fournodes. Very easy to reproduce like you described. 🙏 

I've already submitted a PR to fix it, the difference for your solution is that I used `is_a()` so it checks in the current class, and the parents too. 
check it out here: https://github.com/Laravel-Backpack/CRUD/pull/4390

I will keep this open until @tabacitu merges and tags the release with the fix. 

Thanks again for the report, and for the help here!

Cheers

**Comment by fournodes** [comment]
Hey @pxpm 

May be I am wrong but I do not really see an any advantage to use `is_a` over `instanceof`

**Reasons**
- is_a() is a function, whereas instanceof is a language construct. 
- is_a() function will be significantly slower since it has all the overhead of executing a function call
- they both achieve the same result ? (this is the part I'm unsure) hence attaching an example

```php
class ParentClass {}
class ChildClass extends ParentClass {}
class SubChildClass extends ChildClass {}

$child = new ChildClass();
$subChild = new SubChildClass();

var_dump($child instanceof ParentClass); //true
var_dump(is_a($child, "ParentClass", true)); //true

var_dump($subChild instanceof ParentClass); //true
var_dump(is_a($subChild, "ParentClass", true)); //true

$runs = 1000;

$timer = microtime(true); 
for($i=0; $i<$runs; $i++){
  $child instanceof ParentClass;
}
$instanceOfTime = microtime(true) - $timer;

$timer = microtime(true);
for($i=0; $i<$runs; $i++){
  is_a($child, "ParentClass", true);
}
$isATime = microtime(true) - $timer;

// about 150% slower over just 1000 runs
echo 'Difference: '. round((($isATime - $instanceOfTime) / $instanceOfTime) * 100, 2) . "%"; 
```

**Comment by pxpm** [comment]
Hello @fournodes I was not aware of the performance difference, but 150% in micro-seconds for me still super micro optimization. 

I think it's debatable which one to use and in the end I find `is_a` more readable. I think in this case there is no need for optimization since this would only run once per app cycle, but hey, 150% faster is 150% faster and it probably adds up with other bits of the code. 

I will have a talk with @tabacitu to discuss this in the next week. 

Thanks for raising awareness regarding this. 

PS: already changed the label. 👍

**Comment by tabacitu** [comment]
Merged. Thanks guys! Will be available starting Monday when we release a new patch version.

In regards to the `is_a()` vs `instanceof` debated... I've merged it with `is_a()`. Not because I think it's more readable (I think they're both readable) but because... the performance improvement is so small that it can be approximated to zero. Let's keep the context in mind - we're running on a "fat" framework here, Laravel itself uses `is_a` throughout its code, which gets run upon every request. In my opinion, one more call will not do any harm, on less call will not do any good. Let's not let this distract us from squashing bugs, it really doesn't matter 💪

Thanks again, cheers!
